File: 24  LRU Cache/LRU Cache.js
/**
 * @param {number} capacity
 */
let mp, cache = [], cap;
var LRUCache = function (capacity) {
    mp = new Map(), cache = [];
    cap = capacity;
};

/** 
* @param {number} key
* @return {number}
*/
LRUCache.prototype.get = function (key) {

    if (cache.includes(key)) {
        cache.splice(cache.indexOf(key), 1);
        cache.push(key);
        return mp.get(key);
    }
    else
        return -1;

};

/** 
* @param {number} key 
* @param {number} value
* @return {void}
*/
LRUCache.prototype.put = function (key, value) {
    if (cache.includes(key))
        cache.splice(cache.indexOf(key), 1);
    if (cache.length === cap)
        cache.shift();
    cache.push(key);
    mp.set(key, value);
};

/**
* Your LRUCache object will be instantiated and called as such:
* var obj = new LRUCache(capacity)
* var param_1 = obj.get(key)
* obj.put(key,value)
*/

File: 24  LRU Cache/LRU Cache.cpp
class LRUCache
{
    map<int, int> mp;
    list<int> cache;
    int capacity;

public:
    LRUCache(int capacity)
    {
        this->capacity = capacity;
    }

    int get(int key)
    {
        list<int>::iterator it = find(cache.begin(), cache.end(), key);
        if (it != cache.end())
        {
            cache.erase(it);
            cache.push_back(key);
            return mp[key];
        }
        else
            return -1;
    }
    void put(int key, int value)
    {
        list<int>::iterator it = find(cache.begin(), cache.end(), key);
        if (it != cache.end())
            cache.erase(it);
        if (cache.size() == capacity)
            cache.pop_front();
        cache.push_back(key);
        mp[key] = value;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */